p2p/protocol/ping/ping.go

```diff
@@ -16,88 +16,85 @@ import (
 
 var log = logging.Logger("ping")
 
+// PingSize is the size of ping in bytes
 const PingSize = 32
 
-const ID = "/ipfs/ping/1.0.0"
+// ID is the protocol ID of the ping/2.0.0 protocol
+// This version uses a new stream for every ping.
+const ID = "/ipfs/ping/2.0.0"
 
-const pingTimeout = time.Second * 60
+// IDLegacy is the protocol ID of the ping/1.0.0 protocol
+// This version uses a single stream for all pings.
+const IDLegacy = "/ipfs/ping/1.0.0"
 
 type PingService struct {
 	Host host.Host
 }
 
+// NewPingService creates a new PingService.
+// It uses supports both ping/1.0.0 and ping/2.0.0 for incoming pings
 func NewPingService(h host.Host) *PingService {
 	ps := &PingService{h}
+	h.SetStreamHandler(IDLegacy, ps.PingHandlerLegacy)
 	h.SetStreamHandler(ID, ps.PingHandler)
 	return ps
 }
 
-func (p *PingService) PingHandler(s inet.Stream) {
-	buf := make([]byte, PingSize)
-
-	errCh := make(chan error, 1)
-	defer close(errCh)
-	timer := time.NewTimer(pingTimeout)
-	defer timer.Stop()
-
-	go func() {
-		select {
-		case <-timer.C:
-			log.Debug("ping timeout")
-		case err, ok := <-errCh:
-			if ok {
-				log.Debug(err)
-			} else {
-				log.Error("ping loop failed without error")
-			}
-		}
-		s.Reset()
-	}()
+// PingHandler handles a ping/2.0.0 ping.
+func (ps *PingService) PingHandler(s inet.Stream) {
+	defer s.Close()
 
-	for {
-		_, err := io.ReadFull(s, buf)
-		if err != nil {
-			errCh <- err
-			return
-		}
-
-		_, err = s.Write(buf)
-		if err != nil {
-			errCh <- err
-			return
-		}
+	buf := make([]byte, PingSize)
+	if _, err := io.ReadFull(s, buf); err != nil {
+		log.Debug(err)
+		return
+	}
 
-		timer.Reset(pingTimeout)
+	if _, err := s.Write(buf); err != nil {
+		log.Debug(err)
+		return
 	}
 }
 
+// Ping pings a peer.
+// It first attempts to use ping/2.0.0, and falls back to ping/1.0.0
+// if the peer doesn't support it yet.
 func (ps *PingService) Ping(ctx context.Context, p peer.ID) (<-chan time.Duration, error) {
-	s, err := ps.Host.NewStream(ctx, p, ID)
+	s, err := ps.Host.NewStream(ctx, p, ID, IDLegacy)
 	if err != nil {
 		return nil, err
 	}
+	if s.Protocol() == IDLegacy {
+		return ps.PingLegacy(ctx, p)
```

Comment:
It seems that `PingLegacy` opens a second stream. Is there a reason why we can't use `s`, which has negotiated the legacy protocol already?